orchestrator/relayer/src/valset_relaying.rs

```diff
@@ -21,34 +18,190 @@ use gravity_utils::types::{RelayerConfig, Valset};
+    // order to know exactly which one we must iterate over the history in find_latest_valid_valset()
+
+    let latest_cosmos_valset_nonce = match get_latest_cosmos_valset_nonce(grpc_client).await {
+        Some(n) => n,
+        None => return,
+    };
+
+    // the latest cosmos validator set that it is possible to submit given the constraints
+    // of the validator set currently in the bridge
+    let (latest_submittable_valset, confirms) = match find_latest_valid_valset(
+        latest_cosmos_valset_nonce,
+        &current_valset,
+        grpc_client,
+        &gravity_id,
+    )
+    .await
+    {
+        Ok(v) => v,
+        Err(GravityError::ValsetUpToDate) => return,
+        Err(e) => {
+            error!(
+                "We where unable to find a valid validator set update to submit! {:?}",
+                e
+            );
+            return;
+        }
+    };
+
+    relay_valid_valset(
+        latest_cosmos_valset_nonce,
+        latest_submittable_valset,
+        current_valset,
+        confirms,
+        web3,
+        gravity_contract_address,
+        gravity_id,
+        ethereum_key,
+        timeout,
+        config,
+    )
+    .await;
+}
+
+#[allow(clippy::too_many_arguments)]
+/// determines if the provided validator set should be relayed according to user preferences
+async fn relay_valid_valset(
+    latest_cosmos_valset_nonce: u64,
+    valset_to_relay: Valset,
+    current_valset: Valset,
+    conformations: Vec<ValsetConfirmResponse>,
+    web3: &Web3,
+    gravity_contract_address: EthAddress,
+    gravity_id: String,
+    ethereum_key: EthPrivateKey,
+    timeout: Duration,
+    relayer_config: RelayerConfig,
+) {
+    let cost = ethereum_gravity::valset_update::estimate_valset_cost(
+        &valset_to_relay,
+        &current_valset,
+        &conformations,
+        web3,
+        gravity_contract_address,
+        gravity_id.clone(),
+        ethereum_key,
+    )
+    .await;
+    if cost.is_err() {
+        valset_cost_error(
+            cost,
+            &ethereum_key,
+            &gravity_contract_address,
+            web3,
+            valset_to_relay.clone(),
+            current_valset,
+        )
+        .await;
+        return;
+    }
+    let cost = cost.unwrap();
+
+    info!(
+           "We have detected that valset {} is valid to submit. Latest on Ethereum is {} This update is estimated to cost {} Gas / {:.4} ETH to submit",
+            valset_to_relay.nonce, current_valset.nonce,
+            cost.gas_price.clone(),
+            downcast_to_u128(cost.get_total()).unwrap() as f32
+                / downcast_to_u128(one_eth()).unwrap() as f32
+        );
+
+    let should_relay = match (
+        relayer_config.valset_market_enabled,
+        relayer_config.relay_every_valset,
+    ) {
+        // if the user has configured only profitable relaying then it is our only consideration
+        (true, _) => {
```

Comment:
Considering an "invalid" configuration valset_market_enabled = true, relay_every_valset = true here:

Should we change this config to be an enum with options like "RELAY_ALL_VALSETS", "RELAY_PROFITABLE_VALSETS" and "RELAY_ONLY_NECESSARY_VALSETS"? It deviates from the rest of the configuration options but is clear about what happens.

Otherwise would you consider a config validation step which panics with weird configs? Personally I'd rather have a weird configuration panic than silently do something I don't actually understand, on the other hand it's my responsibility to understand what it's going to do.

Reply:
I was thinking about an enum earlier, I agree that we should make this unambigous, so let me try that.